sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc

```diff
@@ -0,0 +1,202 @@
+supporting this extension must predefine the macro `SYCL_EXT_ONEAPI_FREE_FUNCTION_QUERIES`
+to one of the values defined in the table below. Applications can test for the
+existence of this macro to determine if the implementation supports this
+feature, or applications can test the macro's value to determine which of the
+extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== Overview
+
+The extension allows to get `sycl::id`, `sycl::item`, `sycl::nd_item`,
+`sycl::group` and `sycl::sub_group` instances globally, without having to
+explicitly pass them as arguments to each function used on the device.
+
+NOTE: Passing such instances as arguments can result in a clearer interface
+that is less error-prone to use. For example, if a function accepts a
+`sycl::group`, the caller must assume that function may call a
+`sycl::group_barrier` and ensure that associated control flow requirements are
+satisfied. It is recommended that this extension is used only when modifying
+existing interfaces is not feasible.
+
+== Accessing Instances of Special SYCL Classes
+
+The `sycl::ext::oneapi::this_kernel` namespace contains functionality related
+to the currently executing kernel.
+
+It is the user's responsibility to ensure that that these functions are called
+in a manner that is compatible with the kernel's launch parameters, as detailed
+in the definition of each function. Calling these functions from an incompatible
+kernel results in undefined behavior.
+
+[source,c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace this_kernel {
+
+size_t get_linear_id();
+
+template <int Dimensions>
+id<Dimensions> get_id();
+
+template <int Dimensions>
+item<Dimensions> get_item();
+
+template <int Dimensions>
+nd_item<Dimensions> get_nd_item();
+
+template <int Dimensions>
+group<Dimensions> get_group();
+
+sub_group get_sub_group();
+
+}
+}
+}
+}
+----
+
+[source,c++]
+----
+size_t get_linear_id();
+----
+_Preconditions_: The currently executing kernel must have been launched with a
+`sycl::range` or `sycl::nd_range` argument.
+
+_Returns_: A `size_t` representing the current work-item's linear position in
+the global iteration space. Multi-dimensional indices are converted into a
+linear index following the linearization equations defined in Section 3.11.1 of
+the SYCL 2020 specification.
+
+[source,c++]
+----
+template <int Dimensions>
+id<Dimensions> get_id();
+----
+_Preconditions_: `Dimensions` must match the dimensionality of the currently
+executing kernel. The currently executing kernel must have been launched with a
+`sycl::range` or `sycl::nd_range` argument.
+
+_Returns_: A `sycl::id` instance representing the current work-item in the
+global iteration space.
+
+[source,c++]
+----
+template <int Dimensions>
+item<Dimensions> get_item();
```

Comment:
Sorry, I noticed something else.

The `item` class has a second template parameter `with_offset`.  I think the idea is that you can define your kernel to take `item<Dim, false>` to gain some efficiency if you know it won't be invoked with an offset.  (The `offset` parameter to `parallel_for` is deprecated, but still exists in SYCL 2020.)

The default value of `with_offset` is `true`, so your current specification will always return the "offset" version of `item` even if the user requested the "no offset" version.  It seems like we probably don't want this?  I'm not sure if DPC++ optimizes the "no offset" case currently, but this would be something to check.

An option is to define this free function as always returning the "no offset" case, regardless of how the user defined the kernel.  That might make sense since the "offset" parameter is deprecated.

Reply:
> An option is to define this free function as always returning the "no offset" case, regardless of how the user defined the kernel. That might make sense since the "offset" parameter is deprecated.

I agree this makes sense.  I'll add a non-normative note explaining that because the `offset` parameter is deprecated, we expect that eventually we'll return `item<Dimensions>` and it will mean "no offset".

Reply:
Added something in 055e1578a2845bf8092071f172c320aae835bf69.